Update script to work with released legacy artifacts

jsturdy opened the pull request:
## Description
Script has been tweaked to remove `v2` specifics, as well as pull released artifacts for the legacy code.

### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
Downloads have been tested (OH FW fails because current artifacts aren't in a programmatically accessible location, but this can easily be fixed by the OH FW developers @andrewpeck )

mexanick commented on setup_ctp7.sh:
```diff
@@ -197,94 +198,109 @@ then
-    echo "rm -rf conf.txt"
+    ## Obsolete?
+    set -x
+    mkdir -p vfat3
```
`setLinks.py` script is also required under `vfat3`

jsturdy replied:
is there a technical reason it can't it be used from its current location, i.e., `scripts`?

mexanick replied:
I think that's where the `testConnectivitiy.py` (or one of the scripts it calls) will look for it.

lpetre-ulb replied:
Really? Which script calls it? `replace_parameter.sh` is used in different places, but `setLinks.py` is used only for initial configuration. `grep` does not find `setLinks.py` in `vatqc-python-scripts` nor in `gem-plotting-tools`.

mexanick commented on setup_ctp7.sh:
```diff
@@ -197,94 +198,109 @@ then
+    set -x
+    python ${XHAL_ROOT}/bin/gem_reg.py -n ${ctp7host} \
+           -e update_lmdb ${CARD_GEMDAQ_DIR}/xml/gem_amc_top.xml
+    cp -rfp ${GEM_ADDRESS_TABLE_ROOT}/gem_amc_top.pickle gem_amc_top_v${ctp7fw//./_}.pickle
```
this won't work as is due to different set of libraries used on the host and on the AMC. A `pickle` file has to be specifically altered. See p27 [here](https://github.com/cms-gem-daq-project/sw_utils/blob/develop/v3ElectronicsUserGuide.md#to-update-oh-fw-on-ctp7)

jsturdy replied:
sure, but the file to be altered *is* the one that gets produced here, no?
so it's just an additional manual step after this?

mexanick replied:
yes, exactly. Btw, the host address table has to be updated too **and** the previous pickle file has to be deleted.

jsturdy commented:
> Actually since our legacy release targets compatibility with GE2/1, `recover.sh` should still remain a symlink, as there're differences between GE1/1-V3 and GE2/1 in terms of tx inversion afaik. GE2/1-compatible `recover.sh` file can be found on integration setup (`gem.shelf02.amc07`).

What is supposed to be the source of truth for all these scripts and where is the version controlled source of the most up-to-date files (@evka85?)

Can they be structured in a more coherent/modular way, such that everything that works with `ge11`/`ge21`/`me0` is in a common location, and all specific scripts live in a separate subdirectory?
Or, if there is very little overlap, then just put three releases of the scripts, and a card is only ever set up with one of them at a time (this also matches with what should be the case in the near future)...

lpetre-ulb commented on scripts/initOHv3b.sh:
```diff
@@ -82,7 +82,7 @@ sca.py local $OHMASK program-fpga bit $FILE_FW
 if [[ $FILE_FW =~ (([0-9a-fA-F]+)\.){3}([Bb]\.) ]]; then
     # OHv3b
-    reg_interface.py -e write "GEM_AMC.GEM_SYSTEM.VFAT3.USE_OH_V3B_MAPPING 1"
+    gem_reg.py -e write "GEM_AMC.GEM_SYSTEM.VFAT3.USE_OH_V3B_MAPPING 1"
```
Can't this file be removed altogether? As far as I know, there is no more OHv3a or OHv3b in circulation.

lpetre-ulb commented on scripts/initOHv3b_NoGBTs.sh:
```diff
@@ -39,7 +39,7 @@ sca.py local $OHMASK program-fpga bit $FILE_FW
 if [[ $FILE_FW =~ (([0-9a-fA-F]+)\.){3}([Bb]\.) ]]; then
     # OHv3b
-    reg_interface.py --execute="write GEM_AMC.GEM_SYSTEM.VFAT3.USE_OH_V3B_MAPPING 1"
+    gem_reg.py --execute="write GEM_AMC.GEM_SYSTEM.VFAT3.USE_OH_V3B_MAPPING 1"
```
Same comment for this file.

lpetre-ulb commented on setup_ctp7.sh:
```diff
@@ -59,135 +52,143 @@ shift $((OPTIND-1))
+    usage
 fi
 
+GEM_FW_DIR=/opt/gemdaq/fw
```
Is this variable used anywhere else in the script?

lpetre-ulb commented on setup_ctp7.sh:
```diff
@@ -59,135 +52,143 @@ shift $((OPTIND-1))
 fi
 
+GEM_FW_DIR=/opt/gemdaq/fw
+GEM_ADDRESS_TABLE_ROOT=/opt/cmsgemos/etc/maps
```
Wouldn't it be best to use the environment variable, if it exists?

jsturdy commented:
From `eagle64` I pulled the individual scripts for each GEM version.
What is left is the following:
```
├── blink.sh
├── bugfixes
├── cold_boot_invert_tx.sh
├── cold_boot.sh
├── configure_daq.sh
├── ctp7_status.sh
├── ge11
│   ├── gth_config_opto.sh
│   ├── gth_config_opto_tx_inverted.sh
│   ├── gth_status.sh
│   ├── recover.sh -> recover_v3.sh
│   ├── recover_v2.sh
│   └── recover_v3.sh
├── ge21
│   ├── cold_boot_invert_cxp_rx.sh
│   ├── cold_boot_invert_rx.sh
│   ├── gth_config_opto_cxp_rx_inverted.sh
│   ├── gth_config_opto_rx_inverted.sh
│   ├── gth_config_opto.sh
│   ├── gth_config_opto_TEST.sh
│   ├── gth_config_opto_tx_inverted.sh
│   ├── gth_reset_minipod.sh
│   ├── gth_status.sh
│   ├── recover.sh
│   └── replaceAllParameters.sh
├── generated
│   └── ctp7_status.sh_v1.3.4
├── gth_config_opto.sh
├── gth_config_opto_tx_inverted.sh
├── gth_prbs_cnt_reset.sh
├── gth_prbs_scan.sh
├── gth_prbs_sel.sh
├── gth_reset.sh
├── gth_rx_error_reset.sh
├── gth_status.sh
├── initGBTs.sh
├── initOHv3b_NoGBTs.sh
├── initOHv3b.sh
├── ipb_read.sh
├── ipb_stress_test.sh
├── ipb_write.sh
├── me0
│   ├── cold_boot_invert_cxp_rx.sh
│   ├── cold_boot_invert_rx.sh
│   ├── cold_boot.sh
│   ├── gth_config_opto_cxp_rx_inverted.sh
│   ├── gth_config_opto_rx_inverted.sh
│   ├── gth_config_opto.sh
│   ├── gth_config_opto_TEST.sh
│   ├── gth_config_opto_tx_inverted.sh
│   ├── gth_reset_minipod.sh
│   ├── gth_status.sh
│   ├── recover.sh
│   └── replaceAllParameters.sh
├── read_gem_tests.sh
├── reg_read.sh
├── replace_parameter.sh
├── restart_ipbus.sh
├── sca_test.sh
├── setLinks.py
├── test_clk.sh
├── test.sh
├── ttc_clk_shift.sh
└── warm_boot.sh
```

There are some slight differences between what's in the `ge11` dir and in the `ge21/me0`, but for every file that is in both `ge21` and `me0` there are no differences (so they *could* be moved to the top level)

For the `ge11`/`me0` difference, an example is:
```
for f in $(ls $PWD/*.sh); do fname=${f##*/}; diff $f $PWD/../me0/$fname; echo $fname diff status $?; done
34c34
<               mpoke $((gth_ctrl_ch0_base_addr+256*11)) $data_ch_inverted
---
>               mpoke $gth_ctrl_addr $data_ch_inverted
gth_config_opto.sh diff status 1
34c34
<               mpoke $((gth_ctrl_ch0_base_addr+256*11)) $data_ch_inverted
---
>               mpoke $gth_ctrl_addr $data_ch_inverted
gth_config_opto_tx_inverted.sh diff status 1
13c13
<         for i in `seq 0 35`
---
>         for i in `seq 0 63`
32a33,34
>                 gth_rx_notintable_cnt_base_reg=$(($gth_rx_notintable_cnt_base_reg+256))
>                 gth_rx_disperr_cnt_base_reg=$(($gth_rx_disperr_cnt_base_reg+256))
gth_status.sh diff status 1
10,11c10,11
< echo ./cold_boot_invert_tx.sh
< ./cold_boot_invert_tx.sh
---
> echo ./cold_boot_invert_cxp_rx.sh
> ./cold_boot_invert_cxp_rx.sh
```

@evka85 @lpetre-ulb @mexanick are any of these differences able to be merged (I expect not in the case of `gth_status.sh`, but can't say for the others...)

jsturdy commented:
For differences between `ge21` and `me0` (there are none), the following files exist (can any of them be removed?)
```
for f in $(ls $PWD/*.sh); do fname=${f##*/}; diff $f $PWD/../me0/$fname; echo $fname diff status $?; done
cold_boot_invert_cxp_rx.sh diff status 0
cold_boot_invert_rx.sh diff status 0
gth_config_opto_cxp_rx_inverted.sh diff status 0
gth_config_opto_rx_inverted.sh diff status 0
gth_config_opto.sh diff status 0
gth_config_opto_TEST.sh diff status 0
gth_config_opto_tx_inverted.sh diff status 0
gth_reset_minipod.sh diff status 0
gth_status.sh diff status 0
replaceAllParameters.sh diff status 0
```

lpetre-ulb commented:
> From `eagle64` I pulled the individual scripts for each GEM version.

Is it the authoritative source? I'd say no, but is there a better source...? I'd also probably compare them with the scripts stored in the `GEM_AMC` repo.

> There are some slight differences between what's in the `ge11` dir and in the `ge21/me0`, but for every file that is in both `ge21` and `me0` there are no differences (so they _could_ be moved to the top level)
> 
> For the `ge11`/`me0` difference, an example is:
> 
> ```
> for f in $(ls $PWD/*.sh); do fname=${f##*/}; diff $f $PWD/../me0/$fname; echo $fname diff status $?; done
> 34c34
> <               mpoke $((gth_ctrl_ch0_base_addr+256*11)) $data_ch_inverted
> ---
> >               mpoke $gth_ctrl_addr $data_ch_inverted
> gth_config_opto.sh diff status 1
> 34c34
> <               mpoke $((gth_ctrl_ch0_base_addr+256*11)) $data_ch_inverted
> ---
> >               mpoke $gth_ctrl_addr $data_ch_inverted
> gth_config_opto_tx_inverted.sh diff status 1
> 13c13
> <         for i in `seq 0 35`
> ---
> >         for i in `seq 0 63`
> 32a33,34
> >                 gth_rx_notintable_cnt_base_reg=$(($gth_rx_notintable_cnt_base_reg+256))
> >                 gth_rx_disperr_cnt_base_reg=$(($gth_rx_disperr_cnt_base_reg+256))
> gth_status.sh diff status 1
> 10,11c10,11
> < echo ./cold_boot_invert_tx.sh
> < ./cold_boot_invert_tx.sh
> ---
> > echo ./cold_boot_invert_cxp_rx.sh
> > ./cold_boot_invert_cxp_rx.sh
> ```
> 
> @evka85 @lpetre-ulb @mexanick are any of these differences able to be merged (I expect not in the case of `gth_status.sh`, but can't say for the others...)

The `gth_status.sh` which should be used for all GEM stations is certainly the one which goes up to 63. We are using the MiniPOD transceivers in all cases.

For the other scripts, bugs have been fixed... at some point... on some CTP7... I'd need to cross-check the script; are they available somewhere?

> For differences between `ge21` and `me0` (there are none), the following files exist (can any of them be removed?)

Do we want to support `me0` within the legacy software? While I expect small differences between `ge21` and `me0` in the mid-, long-term, is it something we should care about _now_?

jsturdy commented:
> > From `eagle64` I pulled the individual scripts for each GEM version.
> 
> Is it the authoritative source? I'd say no, but is there a better source...? I'd also probably compare them with the scripts stored in the `GEM_AMC` repo.

This was the point of my above question to @evka85 ...

> 
> > There are some slight differences between what's in the `ge11` dir and in the `ge21/me0`, but for every file that is in both `ge21` and `me0` there are no differences (so they _could_ be moved to the top level)
> > For the `ge11`/`me0` difference, an example is:
> > ```
> > for f in $(ls $PWD/*.sh); do fname=${f##*/}; diff $f $PWD/../me0/$fname; echo $fname diff status $?; done
> > 34c34
> > <               mpoke $((gth_ctrl_ch0_base_addr+256*11)) $data_ch_inverted
> > ---
> > >               mpoke $gth_ctrl_addr $data_ch_inverted
> > gth_config_opto.sh diff status 1
> > 34c34
> > <               mpoke $((gth_ctrl_ch0_base_addr+256*11)) $data_ch_inverted
> > ---
> > >               mpoke $gth_ctrl_addr $data_ch_inverted
> > gth_config_opto_tx_inverted.sh diff status 1
> > 13c13
> > <         for i in `seq 0 35`
> > ---
> > >         for i in `seq 0 63`
> > 32a33,34
> > >                 gth_rx_notintable_cnt_base_reg=$(($gth_rx_notintable_cnt_base_reg+256))
> > >                 gth_rx_disperr_cnt_base_reg=$(($gth_rx_disperr_cnt_base_reg+256))
> > gth_status.sh diff status 1
> > 10,11c10,11
> > < echo ./cold_boot_invert_tx.sh
> > < ./cold_boot_invert_tx.sh
> > ---
> > > echo ./cold_boot_invert_cxp_rx.sh
> > > ./cold_boot_invert_cxp_rx.sh
> > ```
> > 
> > 
> > @evka85 @lpetre-ulb @mexanick are any of these differences able to be merged (I expect not in the case of `gth_status.sh`, but can't say for the others...)
> 
> The `gth_status.sh` which should be used for all GEM stations is certainly the one which goes up to 63. We are using the MiniPOD transceivers in all cases.
> 
> For the other scripts, bugs have been fixed... at some point... on some CTP7... I'd need to cross-check the script; are they available somewhere?

I put the ones I found in this repo, but they should definitely have an authoritative source defined, and **only** reference them from there...

> 
> > For differences between `ge21` and `me0` (there are none), the following files exist (can any of them be removed?)
> 
> Do we want to support `me0` within the legacy software? While I expect small differences between `ge21` and `me0` in the mid-, long-term, is it something we should care about _now_?

I don't care one way or the other, if it works I'd leave it, since it allows to keep everything systematic (until the ansible playbooks are ready)

lpetre-ulb commented:
> > The `gth_status.sh` which should be used for all GEM stations is certainly the one which goes up to 63. We are using the MiniPOD transceivers in all cases.
> > For the other scripts, bugs have been fixed... at some point... on some CTP7... I'd need to cross-check the script; are they available somewhere?
> 
> I put the ones I found in this repo, but they should definitely have an authoritative source defined, and **only** reference them from there...

I guess using the files from this repo is the best that can be done for now...

One problem is that some scripts would better be hosted in `GEM_AMC` (e.g. the GTH related scripts since they are tightly coupled with the firmware) while others should clearly be hosted here  (e.g. `replace_parameter.sh` which is only needed by the legacy software).

> > > For differences between `ge21` and `me0` (there are none), the following files exist (can any of them be removed?)
> > 
> > 
> > Do we want to support `me0` within the legacy software? While I expect small differences between `ge21` and `me0` in the mid-, long-term, is it something we should care about _now_?
> 
> I don't care one way or the other, if it works I'd leave it, since it allows to keep everything systematic (until the ansible playbooks are ready)

:+1:

jsturdy commented:
* last changes I think have been pushed.
* docs site created and updated
  * https://test-gemdaq-ci.web.cern.ch/test-gemdaq-ci/docs/api/gemctp7user/latest/
* added `make_legacy_release.sh` and `setupMachine.sh` (+ helpers)
  * docs have been built with a locally modified version of `make_legacy_release.sh` (to pick up the correct branch for these changes), but the script will work with the "default" legacy branches once changes have been merged in
  * legacy release is mostly standalone, but if you use it to make a new tag, you will have to interact with it
  * the final step, publishing docs and RPMs, requires confirmation

jsturdy commented:
> is it documented somewhere how to actually build the documentation website?

The docs site structure (guides + landing page) are built in the CI of https://gitlab.cern.ch/cms-gem-daq-project/gemdaqsite
Deployment in the CI pipeline is (currently) a manually triggered job by someone with permissions to write to the `cmsgemdaq` EOS project space
The API docs are built in each repo individually (some variables may need to be set, but this is handled transparently by `make_legacy_release.sh`)

AndrewLevin commented:
thanks

if I want to make changes and test them before I make a pull request, is it simple to redirect the build to go to a different url?

jsturdy commented:
> thanks
> 
> if I want to make changes and test them before I make a pull request, is it simple to redirect the build to go to a different url?

Sure, for all of the software repos this is what I used locally:
```
USE_DOXYREST=YES USE_CONDA=YES ENV_NAME=docs-test make doc
```

The only caveat is
* you need `conda` installed, if you build on a CC7 machine with `conda` installed, the above will work, if you obtain `conda` some other way, then you'd have to add the location of the `conda` source script, e.g.,
```
 CONDA_ENV_FILE=/opt/miniconda3/etc/profile.d/conda.sh USE_DOXYREST=YES USE_CONDA=YES ENV_NAME=docs-test make doc
```
then you can browse the built docs from
```
file:///path/to/<repo root>/doc/build/html/index.html
```

If you don't build something that uses doxyrest, you can set up a virtual environment and install `doc/requirements.txt` into that, but make sure `USE_DOXYREST` and `USE_CONDA` aren't set.
Also, this way hasn't been fully tested, as `conda` allowed for it to be completely disposable with non-python dependencies not available via repos
